### PR TITLE
feat: support builderID matching with or without semver for GHA

### PR DESCRIPTION
**Pull request by laurentsimon**

### CODE REVIEW

File: cli/slsa-verifier/main_test.go [hunk 1]
```diff
@@ -513,7 +513,7 @@ func Test_runVerifyArtifactPath(t *testing.T) {
```
Comment (laurentsimon) [inline cli/slsa-verifier/main_test.go:513]: Note: CLI tests on the way

File: cli/slsa-verifier/main_test.go [hunk 2]
```diff
@@ -491,34 +493,82 @@ func Test_runVerifyArtifactPath(t *testing.T) {
...
-					ProvenancePath:      provenancePath,
-					SourceURI:           tt.source,
-					SourceBranch:        tt.pbranch,
-					BuilderID:           tt.pbuilderID,
-					SourceTag:           tt.ptag,
-					SourceVersionTag:    tt.pversiontag,
-					BuildWorkflowInputs: tt.inputs,
+				// TODO(#258): invalid builder ref.
+				semver := path.Base(v)
+				// For each test, we run 4 sub-tests:
```
Comment (asraa) [inline cli/slsa-verifier/main_test.go:498]: For tests whos error expects a build mismatch, this won't (correctly) fail for cases 1-3. 

Maybe you should switch on whether pbuilderID was specified? In those cases we are explicitly testing pBuilderID

Comment (laurentsimon) [reply cli/slsa-verifier/main_test.go:498]: IIUC, only the tests for v0.0.2 use the `@main` builder, the one >= v1 seem to all use a proper `@vx.y.z`, so the build mismatch tests seem to be working. I'll double check.

File: verifiers/internal/gha/verifier.go [hunk 3]
```diff
@@ -59,9 +60,11 @@ func verifyEnvAndCert(env *dsse.Envelope,
 		return nil, nil, err
 	}
 
+	fmt.Println("builderID:", builderID.String())
```
Comment (asraa) [inline verifiers/internal/gha/verifier.go:63]: nit: remove before merging

File: verifiers/internal/gha/verifier.go [hunk 4]
```diff
@@ -147,13 +144,22 @@ func (v *GHAVerifier) VerifyImage(ctx context.Context,
...
-		if verifyErr == nil {
+		if err == nil {
 			return verifiedProvenance, builderID, nil
 		}
+		errs = append(errs, err)
 	}
 
-	return nil, nil, fmt.Errorf("no valid attestations found on OCI registry: %w", verifyErr)
+	// Return the first error.
+	if len(errs) > 0 {
```
Comment (asraa) [inline verifiers/internal/gha/verifier.go:157]: Why do we need this check? If we reach here this means we did not return any provenance, so there must be `errs`?

File: verifiers/utils/builder_test.go [hunk 5]
```diff
@@ -201,6 +202,85 @@ func Test_Matches(t *testing.T) {
...
+			name:      "match long tags",
+			builderID: "some/name@refs/tags/v1.2.3",
+			match:     "some/name@refs/tags/v1.2.3",
+			allowRef:  true,
+		},
+		{
+			name:      "mismatch tag length",
+			builderID: "some/name@refs/tags/v1.2.3",
+			match:     "some/name@v1.2.3",
+			err:       serrors.ErrorMismatchBuilderID,
```
Comment (asraa) [inline verifiers/utils/builder_test.go:276]: sorry if I'm a little lost here, but does a user provide `match` or `builderID`? I thought they provide `match`. Shouldn't it allow user to pass with the short name to provide?

File: verifiers/internal/gha/builder.go [hunk 6]
```diff
@@ -33,63 +34,81 @@ var defaultContainerTrustedReusableWorkflows = map[string]bool{
...
+	// refs/heads/main for e2e tests. See verifyTrustedBuilderRef().
 	// No builder ID provided by user: use the default trusted workflows.
-	if builderID == nil || *builderID == "" {
-		if _, ok := defaultBuilders[path]; !ok {
-			return "", fmt.Errorf("%w: %s got %t", serrors.ErrorUntrustedReusableWorkflow, path, builderID == nil)
+	if expectedBuilderID == nil || *expectedBuilderID == "" {
+		if _, ok := defaultBuilders[certPath]; !ok {
+			return nil, fmt.Errorf("%w: %s got %t", serrors.ErrorUntrustedReusableWorkflow, certPath, expectedBuilderID == nil)
+		}
+		// Construct the builderID using the default builder's name and the certificate's tag.
```
Comment (asraa) [inline verifiers/internal/gha/builder.go:90]: this is not the default right? `certPath` is from the signing certiicate

File: verifiers/internal/gha/provenance.go [hunk 7]
```diff
@@ -42,22 +42,14 @@ func provenanceFromEnv(env *dsselib.Envelope) (prov *intoto.ProvenanceStatement,
...
-	// Note: builderID does not contain the tag.
-	// TODO(#189): support cases where user wants to match on the full builderID, including the tag.
-	bid, err := sourceFromURI(builderID, true)
-	if err != nil {
-		return err
-	}
-	if provid != bid {
+// This function does an exact comparison, and expects certBuilderID to be the full
+// `name@refs/tags/<name>`.
+func verifyBuilderID(prov *intoto.ProvenanceStatement, certBuilderID string) error {
```
Comment (asraa) [inline verifiers/internal/gha/provenance.go:47]: could you change the name to something like `verifyBuilderIDExactMatch`?

**Review by asraa — COMMENTED** [review]
sorry for how long that took! it took me a while to sort things out.

File: verifiers/internal/gha/verifier.go (hunk 3 above)
Comment (laurentsimon) [reply verifiers/internal/gha/verifier.go:63]: good catch. Completely missed this one. Thanks!

File: verifiers/internal/gha/verifier.go (hunk 4 above)
Comment (laurentsimon) [reply verifiers/internal/gha/verifier.go:157]: if there is no attestation, the loop will not enter errs will be nil. I don't know the exact behavior `RunCosignImageVerification`. If it fails when no attestations are found, I agree with you. Shall I keep the code or add a comment that atts is never empty?

File: verifiers/utils/builder_test.go (hunk 5 above)
Comment (laurentsimon) [reply verifiers/utils/builder_test.go:276]: user provides `match`. BuilderID is the one in the provenance. Let me make this clearer in the doc

File: verifiers/internal/gha/builder.go (hunk 6 above)
Comment (laurentsimon) [reply verifiers/internal/gha/builder.go:90]: +1, change made

File: verifiers/internal/gha/provenance.go (hunk 7 above)
Comment (laurentsimon) [reply verifiers/internal/gha/provenance.go:47]: +1, done.

File: verifiers/utils/builder_test.go (hunk 5 above)
Comment (laurentsimon) [reply verifiers/utils/builder_test.go:276]: > Shouldn't it allow user to pass with the short name to provide?

You're correct, but only if `allowRef: true`, which is not set here. The `allowRef` is to ensure we nly enable this feature for GHA builders. The GCB builder just use `name@version`

Comment (laurentsimon) [reply verifiers/utils/builder_test.go]: I renamed to TrustedBuilderID, to make to clear this is the trusted one from the cert / provenance.

File: verifiers/internal/gha/verifier.go (hunk 4 above)
Comment (asraa) [reply verifiers/internal/gha/verifier.go:155]: It is correct that cosign will fail when no attestations are there, but I think we should not rely on that. I overlooked that missing attestations would not enter the loop. Thank you!

**Review by asraa — APPROVED** [review]

File: verifiers/utils/builder_test.go (hunk 5 above)
Comment (laurentsimon) [reply verifiers/utils/builder_test.go]: np!

**Pull request merged**